### ISSUE TITLE
[BUG] ARM64 gcc doesn't support the  `-mbig-endian` flag 

### ISSUE THREAD

**Issue by g-berthiaume** [compiler-explorer/compiler-explorer#2901]
**Problem**   
When adding the `-mbig-endian` flag to the ARM64 gcc compiler (v11)
I get the following error
```
In file included from /opt/compiler-explorer/arm64/gcc-11.1.0/aarch64-unknown-linux-gnu/aarch64-unknown-linux-gnu/sysroot/usr/include/features.h:497,
                 from /opt/compiler-explorer/arm64/gcc-11.1.0/aarch64-unknown-linux-gnu/aarch64-unknown-linux-gnu/sysroot/usr/include/bits/libc-header-start.h:33,
                 from /opt/compiler-explorer/arm64/gcc-11.1.0/aarch64-unknown-linux-gnu/aarch64-unknown-linux-gnu/sysroot/usr/include/stdint.h:26,
                 from /opt/compiler-explorer/arm64/gcc-11.1.0/aarch64-unknown-linux-gnu/lib/gcc/aarch64-unknown-linux-gnu/11.1.0/include/stdint.h:9,
                 from <source>:1:
/opt/compiler-explorer/arm64/gcc-11.1.0/aarch64-unknown-linux-gnu/aarch64-unknown-linux-gnu/sysroot/usr/include/gnu/stubs.h:11:11: fatal error: gnu/stubs-lp64_be.h: No such file or directory
   11 | # include <gnu/stubs-lp64_be.h>
      |           ^~~~~~~~~~~~~~~~~~~~~
compilation terminated.
Compiler returned: 1
```

**How to reproduce**   
You can see by yourself here: https://godbolt.org/z/vTE969837

**Closing note**   
To my understanding of [the official gcc documentation ](https://gcc.gnu.org/onlinedocs/gcc/AArch64-Options.html), the `-mbig-endian` flag should be supported.    
Keep up the good work!

**Comment by Alexhuszagh** [comment]
This should be dependent on the standard library that was compiled with it, even if the front-end supports it, which would require building glibc with big-endian support. Compiling glibc in little-endian mode will produce `include/gnu/stubs-lp64.h`, while compiling in big-endian mode will produce `include/gnu/stubs-lp64_be.h`. `libgcc_s` would also have incompatible symbols.

Since compiler-explorer uses Linaro to download ARM binaries, it would likely require creating a new target, and then downloading releases from [here](http://releases.linaro.org/components/toolchain/binaries/latest-7/aarch64_be-linux-gnu/) to support this target.

**Comment by partouf** [comment]
> This should be dependent on the standard library that was compiled with it, even if the front-end supports it, which would require building glibc with big-endian support. Compiling glibc in little-endian mode will produce `include/gnu/stubs-lp64.h`, while compiling in big-endian mode will produce `include/gnu/stubs-lp64_be.h`. `libgcc_s` would also have incompatible symbols.
> 
> Since compiler-explorer uses Linaro to download ARM binaries, it would likely require creating a new target, and then downloading releases from [here](http://releases.linaro.org/components/toolchain/binaries/latest-7/aarch64_be-linux-gnu/) to support this target.

We no longer use that code that you're referring to.
We build our own cross platform gcc's for the most part: https://github.com/compiler-explorer/gcc-cross-builder

I agree that we would need appropriate special target headers, if you have any tips on how to get those through our cross-builder, please do.

Meanwhile, there are still no "official" ARM builds for gcc 11, so we can't use those instead either.

**Comment by Alexhuszagh** [comment]
> We no longer use that code that you're referring to.

Ah, my mistake.

> We build our own cross platform gcc's for the most part: https://github.com/compiler-explorer/gcc-cross-builder
> 
> I agree that we would need appropriate special target headers, if you have any tips on how to get those through our cross-builder, please do.
> 
> Meanwhile, there are still no "official" ARM builds for gcc 11, so we can't use those instead either.

It looks like you use crosstool-NG, for a project to create cross-compiler toolchains for Docker images, I have a few configurations that would work out-of-the-box, for example, for [aarch64_be-unknown-linux-gnu](https://github.com/Alexhuszagh/xcross/blob/main/ct-ng/arm64eb-unknown-linux-gnu.config). The code is all unlicensed (public domain), so if it's any help, let me know. The config files are auto-generated, and currently only support GCC 10.0, but are trivial to update to newer GCC versions (I'd be happy to test, if needed).

The crucial lines to add to the config would be:

```text
CT_ARCH_DEFAULT_LE=y
CT_ARCH_BE=y
# CT_ARCH_LE is not set
```

**Comment by partouf** [comment]
pinging @TamarChristinaArm do you think this will work?

**Comment by TamarChristinaArm** [comment]
> pinging @TamarChristinaArm do you think this will work?

Yes, but this would needlessly increase the number of compiler configurations. The compiler itself already supports big-endian codegen it's the libc headers that are missing (as mentioned above).

So the easiest/best way is to just enable multilib support in crosstools and specify the big endian target as a multilib.  It looks like crosstools-ng has some support for multilibs so I'll give that a try tomorrow.

> Meanwhile, there are still no "official" ARM builds for gcc 11, so we can't use those instead either.

Coming soon :)

**Comment by mattgodbolt** [comment]
Thanks all!

**Comment by TamarChristinaArm** [comment]
Hmmm so it looks like we don't have any multilibs or multi-arch defined for big-endian inside the source tree.  But looking at it some more, both those options would be overkill even if they were here.

from investigating a bit it looks like glibc only installs the stubs header for the `default-abi` and not all `abi-variants`.  Give me a couple of days to see if I can't get glibc to just support this on there end.

**Comment by TamarChristinaArm** [comment]
I'm testing out a patch to do this, will post soon if it works. should prevent the need to have multiple flavors of toolchains.

**Comment by partouf** [comment]
> This works on trunk now in production. Will recompile 11 soonish

I don't think I ever did this. It works from 11.2 and onwards. Any interest in 11.1 to be recompiled with these fixes?

**Comment by mattgodbolt-molty** [comment]
Thanks for reporting this. We've investigated and found the root cause: the ARM64 GCC 11.1.0 cross-compiler was built without a glibc patch (`001-disable-variants-check.patch`) that's needed for big-endian sysroot support. This patch was added from 11.2.0 onwards in the [gcc-cross-builder](https://github.com/compiler-explorer/gcc-cross-builder) repo, which is why `-mbig-endian` works fine on 11.2+.

Fixing this would require updating the 11.1.0 crosstool-ng build config to apply the same local glibc patch, rebuilding the cross-compiler, and force-reinstalling it on the CE nodes.

Given that 11.2+ all work correctly, we're going to close this as won't-fix for now. If this is still a priority for you and you specifically need 11.1.0 big-endian support, please reopen and we'll get it sorted.

*(I'm Molty, an AI assistant acting on behalf of @mattgodbolt)*

**Comment by g-berthiaume** [comment]
I'm not sure I would trust an LLM with root cause analysis.
That said, I appreciate that it's clearly mentioned that I'm not talking to a human.

I confirmed that the problem has been fixed in newer versions of the compiler.
